ssl: require BoringSSL.

PiotrSikora opened the pull request:
BoringSSL is the TLS library that we develop against and that the build
system uses, so let's not pretend that it works with random versions of
other TLS libraries.

Fixes #759.

PiotrSikora commented on source/common/ssl/openssl.cc:
```diff
@@ -2,39 +2,10 @@
-}
-
-#endif
+void OpenSsl::initialize() { RELEASE_ASSERT(SSL_library_init()); }
```
@mattklein123 let me know if I should move this into `*.h` instead.

mattklein123 replied:
no preference, this is fine.

mattklein123 commented:
LGTM. Let's let this sit for a week or so to see if anyone speaks up.

There are a few doc references that talk about how openssl is still supported. We should fix those to make it clear that boringssl is required.

PiotrSikora commented:
I keep forgetting about the docs. I'll remove references to OpenSSL from there.

btw: should we rename classes from `OpenSsl` to `BoringSsl`?

mattklein123 commented:
Sure, feel free to rename the classes if you like.

PiotrSikora commented:
Last call, any objections before making BoringSSL the only supported choice?

mattklein123 commented:
@PiotrSikora I think we are probably fine to finish this up. Can you merge master and fix the docs? We can merge on Friday.

PiotrSikora commented:
@mattklein123 please take a look.

I replaced the custom code for cipher list verification with new `SSL_CTX_set_strict_cipher_list()` from BoringSSL and added `minLength` constrain, since empty value would be rejected by `SSL_CTX_set_strict_cipher_list()` anyway.

mattklein123 commented on source/common/ssl/openssl.h:
```diff
@@ -4,25 +4,21 @@
- * Global functionality specific to OpenSSL.
+ * Global functionality specific to BoringSSL.
  */
 class OpenSsl final {
```
nit: I think you mentioned this elsewhere but rename to BoringSsl or just something general like "Global" ?

PiotrSikora replied:
Actually, upon further investigation, it looks that we don't need to call `SSL_library_init()` at all, which means that `Ssl::OpenSsl` can be removed.

PiotrSikora replied:
Updated.

PiotrSikora commented on source/common/ssl/context_impl.cc:
```diff
@@ -356,9 +325,7 @@ std::string ContextImpl::getSerialNumber(X509* cert) {
     std::string serial_number(char_serial_number);
     OPENSSL_free(char_serial_number);
-#ifdef OPENSSL_IS_BORINGSSL
     std::transform(serial_number.begin(), serial_number.end(), serial_number.begin(), ::toupper);
```
Note: this is here only for backward-compatibility with existing systems. Does any of your internal systems rely on serial number being uppercase or can we remove this?

mattklein123 replied:
This is only used for admin/debug output. There is no automation. I think we can kill the conversion.

davidben commented:
By the way, BoringSSL provides C++ smart pointers in the header files, so rather than defining your own `CSmartPtr<SSL, SSL_free>`, you can now use `bssl::UniquePtr<SSL>`. They're a bit more efficient too since they don't depend on the two-argument `std::unique_ptr` constructor.

(No need to include a separate header. They're part of the usual headers if we detect C++11 support.)

PiotrSikora commented:
@davidben thanks!

@mattklein123 let me know if I should convert `CSmartPtr<*, *>` to `bssl::UniquePtr<*>` or if you prefer to leave it as-is.

mattklein123 commented:
@PiotrSikora yes please replace if you have time.
